File: Platform/Workspace/Typescript/framework/data/Contains.ts
import { PropertyType, ObjectType } from '@allors/framework';
import { ISessionObject } from '../workspace/ISessionObject';
import { ParameterizablePredicate, ParameterizablePredicateArgs } from './ParameterizablePredicate';
import { CompositeTypes } from '../workspace/Types';

export interface ContainsArgs extends ParameterizablePredicateArgs, Pick<Contains, 'propertyType' | 'object'> {}

export class Contains extends ParameterizablePredicate {
  propertyType: PropertyType;
  object?: CompositeTypes;

  constructor(propertyType: PropertyType);
  constructor(args: ContainsArgs);
  constructor(args: ContainsArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as ContainsArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  public toJSON(): any {
    return {
      kind: 'Contains',
      dependencies: this.dependencies,
      propertyType: this.propertyType.id,
      parameter: this.parameter,
      object: this.object ? ((this.object as ISessionObject).id ? (this.object as ISessionObject).id : this.object) : undefined,
    };
  }
}

File: Platform/Workspace/CSharp/Allors.Workspace/Data/Contains.cs
// <copyright file="Contains.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Contains : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public Contains(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        public IPropertyType PropertyType { get; set; }

        public ISessionObject Object { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                Kind = PredicateKind.Contains,
                Dependencies = this.Dependencies,
                PropertyType = this.PropertyType?.Id,
                Object = this.Object?.Id.ToString(),
                Parameter = this.Parameter,
            };
    }
}
